docs/source/data/fetching-data.mdx

```diff
@@ -1,6 +1,224 @@
+    if (user) {
+      return await this.dbConnection.Booking.findByPk(bookingId);
+    } else {
+      // handle invalid user
+    }
+  }
+
+  //... more methods for finding and creating bookings
+}
+```
+
+</MultiCodeBlock>
+
+> Apollo's [`RestDataSource`](https://github.com/apollographql/datasource-rest/blob/main/src/RESTDataSource.ts) class is a built-out example of how a data source class can handle caching, deduplication, and errors.
+
+## Adding data sources to your context
+
+You can add data sources to your server's `context` initialization function, like so:
+
+<MultiCodeBlock>
+
+```ts title="index.ts"
+//highlight-start
+interface ContextValue {
+  dataSources: {
+    dogsDB: DogsDataSource;
+    catsApi: CatsAPI;
+  };
+  token: string
+}
+//highlight-end
+
+const server = new ApolloServer<ContextValue>({
+  typeDefs,
+  resolvers,
+});
+
+const { url } = await startStandaloneServer(server, {
+  context: async () => {
+    const { cache } = server; // highlight-line
+    const token = req.headers.token;
+    return {
+      // We create new instances of our data sources with each request.
+      // We can pass in our server's cache, contextValue, or any other
+      // info our data sources require. 
+      // highlight-start
+      dataSources: {
+        dogsDB: new DogsDataSource({ cache, contextValue: token }),
+        catsApi: new CatsAPI({ cache }),
+      },
+      //highlight-end
+      token
+    };
+  },
+});
+
+console.log(`🚀  Server ready at ${url}`);
+```
+
+</MultiCodeBlock>
+
+<!-- TODO(AS4) add link to context article once exists  -->
+Apollo Server calls [the `context` initialization](./resolvers/#the-context-argument) function for _every incoming operation_. This means:
+- For every operation, `context` returns an _object_ containing new instances of your data source classes (in this case, `DogsDataSource` and `CatsAPI`).
+- The **`context` function should create a new instance of each data source class for each operation.** You might accidentally combine results from multiple operations if multiple operations share a single data source instance.
+
+Your resolvers can then access your data sources from the shared `contextValue` object and use them to fetch data:
+
+```ts title="resolvers.ts"
+const resolvers = {
+  Query: {
+    dog: async (_, { id }, { dataSources }) => {
+      return dataSources.dogsDB.getDog(id);
+    },
+    popularDogs: async (_, __, { dataSources }) => {
+      return dataSources.dogsDB.getMostLikedDogs();
+    },
+    bigCats: async (_, __, { dataSources }) => {
+      return dataSources.catsApi.getCats({ size: 10 });
+    },
+  },
+};
+```
+
+## Open-source implementations
+
+[In Apollo Server 3](/apollo-server/v3/data/data-sources), you'd subclass the `DataSource` abstract class for each of your data sources. In Apollo Server 4, we recommend [making a custom class for each data source](#creating-data-source-classes), with each class best suited for that particular source of data.
```

Comment:
Maybe something like:

```suggestion
Apollo Server 3 contained [a small abstract class named `DataSource`](/apollo-server/v3/data/data-sources) that you could subclass for each of your data sources. Data sources were initialized in a special `dataSources` function instead of the main `context` function. In Apollo Server 4, you can create your data sources in the same `context` function as the rest of your per-request setup, and there is no special `DataSource` superclass. We recommend [making a custom class for each data source](#creating-data-source-classes), with each class best suited for that particular source of data.
```

Reply:
I like the feeling behind this! I did a few minor tweaks for style:
```suggestion
Apollo Server 3 contained [an abstract class named `DataSource`](/apollo-server/v3/data/data-sources) that each of your data sources could subclass. You'd then initialize each of your `DataSource` subclasses using a special `dataSources` function, which would attach your data sources to your `context` behind the scenes. 

In Apollo Server 4, you can now create your data sources in the same `context` function as the rest of your per-request setup, avoiding the `DataSource` superclass entirely. We recommend [making a custom class for each data source](#creating-data-source-classes), with each class best suited for that particular source of data.
```

Reply:
Looks good